twoftpd-auth.1
.TH twoftpd-auth 1
.SH NAME
twoftpd-auth \- TwoFTPD authentication front-end
.SH SYNOPSIS
.B twoftpd-auth
.B cvmodule
.B twoftpd-xfer
.SH DESCRIPTION
.B twoftpd-auth
reads a username and password from the network and authenticates them
using the CVM module
.BR cvmodule .
If the authentication succeeds, it sets the environment variables
.BR UID ,
.BR GID ,
.BR HOME ,
and
.B USER
to the user's UID, GID, home directory, and login name respectively,
and then executes
.BR twoftpd-xfer .
It also sets
.B GROUP
if the CVM reports a group name, and
.B AUTHENTICATED
to
.B 1
.RB ( twoftpd-auth
and
.B twoftpd-xfer
look for this variable to determine what startup banner code to use).
.SH ENVIRONMENT
.TP
.B AUTH_ATTEMPTS
The maximum number of authentication attempts to allow before dropping
the connection.
.TP
.B AUTH_TIMEOUT
The maximum number of seconds to wait for successful authentication.
.TP
.B LOGINBANNER
If set, the contents of this variable will be sent as extra responses
before the user logs in.
.TP
.B TIMEOUT
The maximum number of seconds to wait for command input.
.SH SEE ALSO
twoftpd-xfer(1)


NEWS
-------------------------------------------------------------------------------
Changes in version 1.30

- Switched to the CVM v2 client library.

- Added a $AUTH_ATTEMPTS control to twoftpd-auth to limit the number of
  authentication attempts per session.

- Rewrote the directory listing code to fix listing files with special
  characters in their names.

- Added a transfer rate (bytes/sec) indicator to transfer logs.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 1.21

- Added a fake listing routine to twoftpd-drop, to avoid errors on GUI
  clients that require LIST or NLST.

- Fixed a compile failure on FreeBSD in the newer MDTM command.
  Thanks Sergei Kolobov.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 1.20

- Modified the backends (twoftpd-auth and twoftpd-xfer) to show a 220
  banner if the front end authenticated, and 230 if not.  This fixes
  problems some FTP clients have when running twoftpd-anon after
  authentication.

- Added twoftpd-drop, a write-only FTP server.

- Added twoftpd-switch, a tool to choose between twoftpd-anon,
  twoftpd-drop, and twoftpd-xfer based on the value of $SHELL,
  $SHELL_READONLY, and $SHELL_WRITEONLY.

- Make sure to open only regular files.

- Added a MDTM command to the xfer backend that can set timestamps
  (instead of just retrieving them).

- Fixed bug exposed by Mozilla which tries to retrieve directories
  before listing them.

- Eliminated the bogus APPE-after-REST test.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 1.17

- Fixed a bug where paths starting with "/." would not be omitted when
  nodotfiles was set.
- Check the return value from *buf_close functions.
- Fixed a minor (but annoying) typo in listing permissions of sticky
  files.
- Path permission failures (no dot files / lock home) now produce a
  sensible system error.
- Fixed a bug where listing "/foo/*" would not include the "/foo/"
  prefix in the resulting names.
- Updated for bglibs version 1.006

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 1.16

- Added a port-20 socket hack, based partly on code contributed by
  Wejn.
- Fixed problems with handling interrupted system calls (strace would
  kill twoftpd).
- Converted to use bglibs.
- Added support for restarting uploads with REST.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 1.15

- Remember to uncork the socket before closing it.

- Fixed a bug in twoftpd-anon-conf -- getpwnam returns the same static
  struct each time it's called, but I used it in two different ways.

- Fixed a minor bug in CWD command to make it possible to cd into
  execute-only directories.

- Fixed handling when doing NLST against non-existant files.

- When making/accepting connections, abort if stdin becomes readable
  (because of additional command or hangup).

- Added support for handling supplemental group IDs.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 1.14

- Fixed a bug in the string sort implementation that caused listings to
  leak memory.

- Fixed up PASV connection mode to only show the 150 response if the
  connection does not fail.

- Some other minor fixes to socket handling and response messages.

- Added a work around for broken firewalls that rewrite certain commands
  (specifically PORT) with two CR characters before the final LF.

- Fixed missing socket library for SysV (ie Solaris) systems.

- Strip trailing slashes, which confuse at least the "lock $HOME" mode,
  from the home path.

- Use corked (Linux) or no-push (BSD) writes to improve transfer
  efficiency.

- Set SO_LINGER on outbound sockets (LIST/RETR) to prevent the success
  message from being sent on the network before the data is completely
  sent.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 1.13

- Bugfix: listing "." when CWD is "/" would fail.

- Include system error message in STAT and MDTM.

- Print out an error message when the session or command times out,
  instead of just silently dropping the connection.

- Modified the PASV response to work around yet more broken clients.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 1.12

- Now uses the CVM 0.10 (protocol 1) API for authentication.  This means
  you MUST use modules from CVM 0.10 or later with this version.

- Many error messages now include the system error message (like "File
  or path not found") that caused the error.

- Added a simple, limited LIST/NLST option parser, to provide
  compatibility with broken clients that like to send commands like
  "NLST -alL".

- Added authentication and session timeout limits.

- SITE commands are no longer output twice in the logs.

Development of this version has been sponsored by FutureQuest, Inc.

twoftpd-auth.c
/* twoftpd-auth.c - Authentication front-end for twoftpd
 * Copyright (C) 2008  Bruce Guenter <bruce@untroubled.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <sysdeps.h>
#include <cvm/v2client.h>

#include "twoftpd.h"

const char program[] = "twoftpd-auth";

static char** argv_xfer = 0;

static const char* user = 0;
static const char* cvmodule = 0;
static unsigned long auth_attempts;

static void do_exec()
{
  alarm(0);
  if (putenv("AUTHENTICATED=1") == 0
      && cvm_setenv())
    execvp(argv_xfer[0], argv_xfer);
  respond(421, 1, "Could not execute back-end.");
  exit(1);
}

static int handle_user(void)
{
  if (user) free((char*)user);
  user = strdup(req_param);
  return respond(331, 1, "Send PASS.");
}

static int handle_pass(void)
{
  if (!user) return respond(503, 1, "Send USER first.");
  if (cvm_authenticate_password(cvmodule, user, getenv("TCPLOCALHOST"),
				req_param, 0) == 0)
    do_exec();
  free((char*)user);
  user = 0;
  if (!respond(530, 1, "Authentication failed."))
    return 0;
  if (auth_attempts-- <= 1) {
    respond(421, 1, "Too many authentication failures.");
    return 0;
  }
  return 1;
}

const command verbs[] = {
  { "USER", 0, 0, handle_user },
  { "PASS", 1, 0, handle_pass },
  { 0,      0, 0, 0 }
};

const command site_commands[] = {
  { 0,      0, 0, 0 }
};

int startup(int argc, char* argv[])
{
  const char* tmp;
  unsigned long auth_timeout;
  
  if (argc < 3) {
    respond(421, 1, "Configuration error, insufficient paramenters.");
    return 0;
  }
  
  cvmodule = argv[1];
  argv_xfer = argv + 2;

  if (!getenv("SERVICE") && putenv("SERVICE=ftp") == -1) {
    respond(421, 1, "Error setting $SERVICE.");
    return 0;
  }

  if ((tmp = getenv("LOGINBANNER")) != 0) show_banner(220, tmp);

  auth_timeout = 0;
  if ((tmp = getenv("AUTH_TIMEOUT")) != 0) auth_timeout = strtou(tmp, &tmp);
  alarm(auth_timeout);

  auth_attempts = ~0UL;
  if ((tmp = getenv("AUTH_ATTEMPTS")) != 0)
    auth_attempts = strtou(tmp, &tmp);

  return respond(220, 0, "TwoFTPd server ready.") &&
    respond(220, 1, "Authenticate first.");
}
